include/envoy/upstream/upstream.h

```diff
@@ -857,6 +857,12 @@ class ClusterInfo {
    */
   virtual void createNetworkFilterChain(Network::Connection& connection) const PURE;
 
+  /**
```

Comment:
Does this need to be an interface function? Could it just be a utility function? I don't feel that strongly about it but throwing it out there.

Reply:
The HTTP protocol decision takes features (which inferred by http2_protocol_settings and protocol_selection fields from cluster.proto) from cluster info, so that's why I put here, (alternatives considered in decoder callbacks or utility).

Reply:
My guess is the logic will need change with HTTP3 support and takes more cluster info (obviously use downstream protocol doesn't work on QUIC cluster and vice versa)

Reply:
OK sounds good.